Project Exercise 1

shiitake opened the pull request:
### PROJECT EXERCISE
Your Product Team has just started the next sprint, working on a new Pharmacy management system.
You pull the next feature story from your team’s backlog, assign it to yourself, and get to work on the following criteria:
* The business needs an API that can update a single Pharmacy record. 
* The user can update any number of fields.
* The user can select any one pharmacy to update, from a list of 5 existing pharmacies in the database. 

 Each Pharmacy record has the following fields:

* Name
* Address
* City
* State
* Zip
* Number of Filled Prescriptions (current month to date)
* Created Date (date pharmacy was added to the system)
* Updated Date (date pharmacy was last updated in the system)
 
### SOLUTION

You must create a new C# .Net 7 Web API solution that contains the following:
* Pharmacy API Controller that allows the following:
  * View a Pharmacy By ID
  * View the entire Pharmacy List – all 5 pharmacies
  * Update a Pharmacy By ID
* Business Service
  * Provides logic and actions that support the Controller methods
* Repository
  * Provides the Business Service with read/write access to the database tables
* MS SQL Database (SQLExpress is fine)
  * Contains the tables necessary to support the entities mentioned above
  * Tables contain the necessary data / records to support this exercise’s requirements
 
### BONUS OBJECTIVES
* Use a dependency injection pattern – leveraging interfaces and initial configuration injection
* Use standard authentication, authorization and/or routing attributes
* Setup of a SQL database using code-first migrations.
* Incorporate Entity Framework
* Create a UI that leverages these functions
* Unit Test coverage - OR - Automated API Test Coverage (ex: using a Jasmine-like test harness).

shiitake committed:
Initial commit

shiitake committed:
moved api to server directory

shiitake committed:
Added unit tests

shiitake committed:
Added database instructions

Amuuo commented on src/server/HelloCodeaholics.Web/Program.cs:
```diff
@@ -0,0 +1,36 @@
+builder.Services.AddSwaggerGen();
+
+var connectionString = builder.Configuration.GetConnectionString("HelloCodeEntities");
+builder.Services.AddDbContext<HelloCodeDbContext>(options => options.UseSqlServer(connectionString));
```
something worth checking out: [DbContextPool](https://learn.microsoft.com/en-us/ef/core/performance/advanced-performance-topics?tabs=with-di%2Cexpression-api-with-constant)

shiitake replied:
Will check it out.  Thanks.

Amuuo commented on src/server/HelloCodeaholics.Services/Application/PharmacyService.cs:
```diff
@@ -0,0 +1,45 @@
+using HelloCodeaholics.Infrastructure;
+using Microsoft.EntityFrameworkCore;
+
+namespace HelloCodeaholics.Services.Application
+{
+    public class PharmacyService : IPharmacyService
+    {
+        private readonly HelloCodeDbContext _context;
+
+        public PharmacyService(HelloCodeDbContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<Pharmacy?> GetPharmacyById(int id)
+        {
+            return await _context.Pharmacies.Where(x => x.Id == id).FirstOrDefaultAsync();
+        }
+
+        public async Task<List<Pharmacy>> GetPharmacyList()
+        {
+            return await _context.Pharmacies.ToListAsync();
+        }
+
+        public async Task<bool> PharmacyExists(int id)
+        {
+            return await _context.Pharmacies.AnyAsync(x => x.Id == id);
+        }
+        public async Task<Pharmacy> UpdatePharmacy(Pharmacy pharmacy)
+        {
+            pharmacy.UpdatedDate = DateTime.Now;
+            _context.Update(pharmacy);
+            await _context.SaveChangesAsync();
+            return pharmacy;
+        }
+    }
+}
```
Couple new C# features that may or may not be on your radar:

[Primary Constructors](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-12.0/primary-constructors)

Simplified namespace declaration: 
![image](https://github.com/shiitake/hello-codeaholics/assets/30686676/48c4d36a-aa37-4555-af54-cef72ec02d42)

Doesn't change anything logically but the code looks a little bit cleaner:

![image](https://github.com/shiitake/hello-codeaholics/assets/30686676/700b15bf-4b83-40c1-9084-b776f1843095)

shiitake replied:
I agree.

Amuuo commented on src/server/HelloCodeaholics.Web/Controllers/PharmacyController.cs:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        [HttpGet]
+        public async Task<ActionResult<List<Pharmacy>>> GetPharmacies()
```
Curious why you chose the name `GetPharmacies` for the controller method, and `GetPharmacyList` for the service method?

shiitake replied:
Hmmm. No burning reason either way.  

It's easy enough to rename `GetPharmacyList` to `GetPharmacies` to provide some consistancy.

DeryGermann replied:
Hi Shannon, this is Dery and I work alongside Adam and Jyoti. _(Just an FYI)_

I would argue towards naming the method `GetPharmacyList` instead of `GetPharmacies` as it seems to be more descriptive and less ambiguous. It will also help those who are not proficient in English to know exactly what that method is doing without having to drill into said method.

shiitake replied:
Done.

shiitake committed:
Updates per PR feedback

shiitake committed:
Renamed Projects

Added Exception Handling

Removed NotFound status codes.

GET responses now utilize AsNoTracking

Renamed GetPharmacies to GetPharmacyList

Added pagination options for GetPharmacyList

Change Pharmacy update from PUT to POST, returning updated entity

Added CreatedBy and UpdatedBy to Pharmacy entities and created new migration

Set CreatedDate and CreatedBy to IsModified = false when updating

Moved Pharmacy validation on Update to PharmacyService

Added UnitTest for Pharmacy validation

shiitake commented:
Per the feedback session yesterday I've made the following changes:

- Added Exception Handling
- Removed `NotFound` status codes.
- GET responses now utilize `AsNoTracking()`
- Renamed `GetPharmacies` to `GetPharmacyList`
- Added pagination options for `GetPharmacyList`
- Change Pharmacy update from **PUT** to **POST**, returning updated entity
- Added `CreatedBy` and `UpdatedBy` to Pharmacy entities and created new migration
- Set `CreatedDate` and `CreatedBy` to `IsModified = false` when updating
- Moved Pharmacy validation on Update to PharmacyService
- Added UnitTest for Pharmacy validation

Please let me know if you see anything else @Amuuo @DeryGermann

shiitake committed:
Created repository for database data access

shiitake committed:
create dto and mappers
